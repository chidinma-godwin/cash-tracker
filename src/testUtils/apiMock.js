import {
  loginEndpoint,
  signupEndpoint,
  transactionEndpoint,
  userEndpoint,
} from 'constants/endpoints';
import { createServer, Model } from 'miragejs';

export default function makeServer(environment = 'test') {
  return createServer({
    environment,
    models: {
      user: Model,
      transaction: Model,
    },
    routes() {
      this.get(userEndpoint, () => ({
        user: {
          id: '1',
          username: 'Authenticated User',
          email: 'user@email.com',
          clientsEmail: ['client1@email.com'],
          pendingInvitations: ['invite@email.com'],
          pendingRequests: ['pending1@email.com', 'pending2@email.com'],
        },
      }));
      this.post(signupEndpoint, (schema, request) => {
        const attrs = JSON.parse(request.requestBody);
        return schema.users.create(attrs);
      });
      this.post(loginEndpoint, (schema, request) => {
        const { email, password } = JSON.parse(request.requestBody);
        return schema.users.findBy({ email, password });
      });
      this.get(`${transactionEndpoint}/:id`, (schema, request) => {
        const { id } = request.params;
        return schema.transactions.where(transaction =>
          transaction.participants.includes(id)
        );
      });
      this.post(transactionEndpoint, (schema, request) => {
        const attrs = JSON.parse(request.requestBody);
        return schema.transactions.create(attrs);
      });
    },
  });
}
