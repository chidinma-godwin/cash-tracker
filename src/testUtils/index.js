import React from 'react';
import { SWRConfig } from 'swr';
import { render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';

// eslint-disable-next-line react/prop-types
const Wrapper = ({ children }) => (
  <SWRConfig value={{ dedupingInterval: 0 }}>{children}</SWRConfig>
);

const customRender = (ui, options) =>
  render(ui, { wrapper: Wrapper, ...options });

const customRenderHook = hook => renderHook(hook, { wrapper: Wrapper });

export * from '@testing-library/react';
export { customRenderHook as renderHook };
export { customRender as render };
