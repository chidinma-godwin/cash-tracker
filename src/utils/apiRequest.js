import {
  unexpected,
  wrongInput,
  notFound,
  unauthenticated,
} from 'constants/errorMessages';

const apiRequest = async (url, body) => {
  let err = '';
  const response = await fetch(url, {
    method: body ? 'POST' : 'GET',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(body),
  });
  if (response.status === 400) {
    err = wrongInput;
  } else if (response.status === 500) {
    err = unexpected;
  } else if (response.status === 404) {
    err = notFound;
  } else if (response.status === 401) {
    err = unauthenticated;
  } else {
    err = unexpected;
  }
  return { response, err };
};

export default apiRequest;
