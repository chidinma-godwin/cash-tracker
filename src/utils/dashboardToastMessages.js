export const getNotModifiedErrMsg = ({ action, type, setErrMsg }) => {
  if (action === 'add' && type === 'request') {
    setErrMsg('You have already sent a connection request to this user');
  } else if (action === 'add') {
    setErrMsg('You already have an account with this user');
  } else if (action === 'delete' && type === 'request') {
    setErrMsg('You have already cancelled your connection request');
  } else if (action === 'delete' && type === 'account') {
    setErrMsg('You have already deleted your account with this user');
  } else {
    setErrMsg('You have already declined connection invitation from this user');
  }
};

export const getUpdateSuccessfulMessage = ({ action, type, setSuccessMsg }) => {
  if (action === 'add' && type === 'request') {
    setSuccessMsg('Invitation sent successfully');
  } else if (action === 'add') {
    setSuccessMsg('Transaction account successfully confirmed');
  } else if (action === 'delete' && type === 'request') {
    setSuccessMsg('Transaction account request cancelled');
  } else if (action === 'delete' && type === 'account') {
    setSuccessMsg('Transaction account deleted successfully');
  } else {
    setSuccessMsg('Transaction account invitation declined');
  }
};
export const getUpdateFailedMsg = ({ action, type, setErrMsg }) => {
  if (action === 'add' && type === 'request') {
    setErrMsg('Unable to invite this user. Please try again.');
  } else if (action === 'add') {
    setErrMsg('Unable to confirm this account. Please try again.');
  } else if (action === 'delete' && type === 'request') {
    setErrMsg('Unable to cancel request. Please try again.');
  } else if (action === 'delete' && type === 'account') {
    setErrMsg(
      'Unable to delete your account with this user. Please try again.'
    );
  } else {
    setErrMsg('Unable to decline this invitation. Please try again.');
  }
};
