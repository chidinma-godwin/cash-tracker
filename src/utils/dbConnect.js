import mongoose from 'mongoose';

const connection = {};

async function dbConnect() {
  if (connection.isConnected) {
    return;
  }

  /* Connecting to our database */
  const db = await mongoose.connect(process.env.MONGODB_URI, {
    useNewUrlParser: true,
    useUnifiedTopology: true,
    useFindAndModify: false,
    useCreateIndex: true,
  });
  console.log('database connected');
  connection.isConnected = db.connections[0].readyState;
}

export default dbConnect;
