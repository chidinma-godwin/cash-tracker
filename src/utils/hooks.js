import useSwr from 'swr';
import fetch from 'isomorphic-unfetch';

export const fetcher = url => fetch(url).then(response => response.json());

export const useUser = () => {
  const { data, error, mutate } = useSwr('/api/user', fetcher);
  const loading = !data && !error;
  const user = data?.user;
  return { user, loading, mutate };
};
