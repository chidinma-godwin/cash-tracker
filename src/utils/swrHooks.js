import useSwr from 'swr';

// CashTracker imports
import { userEndpoint, transactionEndpoint } from 'constants/endpoints';

export const fetcher = async url => {
  const res = await fetch(url);
  if (!res.ok) {
    const error = new Error();
    error.status = res.status;
    throw error;
  }
  return res.json();
};

export const useUser = () => {
  const { data, error, mutate } = useSwr(userEndpoint, fetcher);
  const loading = !data && !error;
  const user = data?.user;
  return { user, loading, mutate };
};

export const useTransaction = id => {
  const { data, error, mutate } = useSwr(
    `${transactionEndpoint}/${id}`,
    fetcher
  );
  const loading = !data && !error;
  const transactions = data?.transactions;
  return { transactions, loading, mutate };
};
