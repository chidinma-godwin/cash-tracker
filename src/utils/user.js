// CashTracker Imports
import User from 'models/User';

export async function createUser({ name, email, password, password2 }) {
  const user = await User.findOne({ email });
  if (user) {
    return new Error('A user with this email already exist');
  }
  if (password !== password2) {
    return new Error('Password do not match!');
  }
  const pendingInvitationsUsers = await User.find({ pendingRequests: email });
  const pendingInvitationsEmail =
    pendingInvitationsUsers &&
    pendingInvitationsUsers.map(pendingUser => pendingUser.email);
  const newUser = await User.create({
    username: name,
    email,
    password,
    ...(pendingInvitationsEmail && {
      pendingInvitations: pendingInvitationsEmail,
    }),
  });
  return { username: newUser.username, email: newUser.email };
}

export async function findUser(email, password) {
  const currentUser = await User.findOne({ email });
  if (currentUser && (await currentUser.passwordMatch(password))) {
    delete currentUser.password;
    return currentUser;
  }
  return new Error('Incorrect login credentials');
}
