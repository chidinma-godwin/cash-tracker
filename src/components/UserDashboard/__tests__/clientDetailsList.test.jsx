import singletonRouter from 'next/router';
import userEvent from '@testing-library/user-event';

// CahsTracker imports
import { render, screen } from 'testUtils';
import { useUser } from 'utils/hooks';
import ClientDetailsList from '../clientDetailsLit';

jest.mock('utils/hooks', () => ({
  useUser: jest.fn(),
}));

// eslint-disable-next-line global-require
jest.mock('next/dist/client/router', () => require('next-router-mock'));

describe('ClientDetailsList component', () => {
  const requiredProps = {
    handleTransactionAccount: jest.fn(),
    handleShowDialog: jest.fn(),
  };
  beforeEach(() => {
    useUser.mockReturnValue({
      user: {
        clientsEmail: ['client1@email.com', 'client2@email.com'],
        clientsDetails: [
          { username: 'client1', email: 'client1@email.com' },
          { username: 'client2', email: 'client2@email.com' },
        ],
        deletedEmails: ['deleted@email.com'],
        pendingRequests: ['pending@email.com'],
      },
    });
  });
  it('matches snapshot', () => {
    const { asFragment } = render(<ClientDetailsList {...requiredProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('should navigate to the correct url if the view button of a transaction account is clicked', () => {
    render(<ClientDetailsList {...requiredProps} />);
    window.scrollTo = jest.fn();

    const viewButton = screen.getAllByRole('link', {
      name: /View transaction/i,
    });
    userEvent.click(viewButton[0]);

    expect(singletonRouter).toMatchObject({
      asPath: '/transaction-details?clientEmail=client1%40email.com',
    });
  });
  it('calls handleTransactionAccount props with the right arguments if the delete button is clicked', () => {
    render(<ClientDetailsList {...requiredProps} />);

    const deleteButton = screen.getAllByRole('button', {
      name: /delete account/i,
    });
    userEvent.click(deleteButton[0]);

    expect(requiredProps.handleTransactionAccount).toHaveBeenCalledWith(
      'client1@email.com',
      'delete',
      'account'
    );
  });
  it('calls handleShowDialog props when the add new account button is clicked', () => {
    render(<ClientDetailsList {...requiredProps} />);

    const addButton = screen.getByRole('button', { name: 'Add New Account' });
    userEvent.click(addButton);

    expect(requiredProps.handleShowDialog).toHaveBeenCalled();
  });
  it('displays correct info, when the user has no clients', () => {
    useUser.mockReturnValue({
      user: {
        clientsEmail: [],
        clientsDetails: [],
        deletedEmails: [],
        pendingRequests: [],
      },
    });
    render(<ClientDetailsList {...requiredProps} />);

    expect(
      screen.getByText(/You do not have any transaction account/)
    ).toBeInTheDocument();
    expect(
      screen.getByRole('button', { name: 'Add New Account' })
    ).toBeInTheDocument();
  });
});
