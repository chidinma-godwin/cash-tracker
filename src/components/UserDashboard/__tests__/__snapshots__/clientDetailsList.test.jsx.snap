// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClientDetailsList component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root MuiPaper-elevation8 MuiPaper-rounded"
  >
    <div
      class="MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-content"
      >
        <span
          class="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
        >
          Transaction Accounts
        </span>
        <span
          class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body1 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
        >
          Username of people you have a transaction account with
        </span>
      </div>
    </div>
    <div
      class="MuiCardContent-root"
    >
      <ul
        class="MuiList-root MuiList-padding"
      >
        <li
          class="MuiListItem-container"
        >
          <div
            class="MuiListItem-root MuiListItem-gutters MuiListItem-secondaryAction"
          >
            <div
              class="MuiListItemAvatar-root"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circle MuiAvatar-colorDefault"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                  />
                </svg>
              </div>
            </div>
            <div
              class="MuiListItemText-root MuiListItemText-multiline"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
              >
                client1
              </span>
              <p
                class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                client1@email.com
              </p>
            </div>
          </div>
          <div
            class="MuiListItemSecondaryAction-root"
          >
            <a
              aria-disabled="false"
              aria-label="view transaction"
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
              edge="end"
              href="/transaction-details?clientEmail=client1%40email.com"
              tabindex="0"
            >
              <span
                class="MuiButton-label"
              >
                View
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </a>
            <button
              aria-label="delete account"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </li>
        <li
          class="MuiListItem-container"
        >
          <div
            class="MuiListItem-root MuiListItem-gutters MuiListItem-secondaryAction"
          >
            <div
              class="MuiListItemAvatar-root"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circle MuiAvatar-colorDefault"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                  />
                </svg>
              </div>
            </div>
            <div
              class="MuiListItemText-root MuiListItemText-multiline"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
              >
                client2
              </span>
              <p
                class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                client2@email.com
              </p>
            </div>
          </div>
          <div
            class="MuiListItemSecondaryAction-root"
          >
            <a
              aria-disabled="false"
              aria-label="view transaction"
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
              edge="end"
              href="/transaction-details?clientEmail=client2%40email.com"
              tabindex="0"
            >
              <span
                class="MuiButton-label"
              >
                View
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </a>
            <button
              aria-label="delete account"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </li>
      </ul>
    </div>
    <div
      class="MuiCardActions-root MuiCardActions-spacing"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSizeSmall MuiButton-sizeSmall"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Add New Account
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
